### PR TITLE
SNOW-1618623 lazy index: refactor qc as input for index constructor

### PR DESCRIPTION
**Pull request by sfc-gh-azhan**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1618623

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

lazy index: 

- refactor qc as input for index constructor and make it consistent with other init methos like dataframe and series.
- raise ValueError for `pd.Index(df)`

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 1]
```diff
@@ -97,12 +95,15 @@ def __init__(
...
         """
-        self._parent = data if isinstance(data, BasePandasDataset) else None
-        data = data._query_compiler if isinstance(data, BasePandasDataset) else data
-        if isinstance(data, SnowflakeQueryCompiler):
-            qc = data
+        if query_compiler:
+            self._query_compiler = query_compiler
+        elif isinstance(data, BasePandasDataset):
+            self._parent = data
+            self._query_compiler = data._query_compiler.drop(
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:102]: I know this behavior existed before this PR, but it doesn't match pandas:

```python
df = pd.DataFrame([1, 2])
# in pandas, get ValueError: Index data must be 1-dimensional
# in Snowpark pandas, get Index([0, 1], dtype='int64')
print(pd.Index([1, 2]))
```

Is this behavior intentional? Do we document it?

File: CHANGELOG.md [hunk 2]
```diff
@@ -75,6 +75,7 @@
 
 ### Behavior change
 - `Dataframe.columns` now returns native pandas Index object instead of Snowpark Index object.
+- Refactor and introduce `query_compiler` argument in `Index` constructor to create `Index` from query compiler.
```
Comment (sfc-gh-mvashishtha) [inline CHANGELOG.md:78]: Users shouldn't be aware of the `query_compiler` argument to the `Index` constructor. Should we be noting this change at all in the release notes?

Comment (sfc-gh-azhan) [reply CHANGELOG.md:78]: why? the constructor is a public API though.

File: src/snowflake/snowpark/modin/plugin/extensions/index.py (hunk 1 above)
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py:102]: I cannot follow the above examples they should both work.

Comment (sfc-gh-mvashishtha) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py:102]: Sorry, I edited my example. I meant constructing an index from the dataframe.

File: CHANGELOG.md (hunk 2 above)
Comment (sfc-gh-mvashishtha) [reply CHANGELOG.md:78]: It's a public API, but we also don't want or expect users to construct indexes out of query compilers at all.

I would consider this change a refactor that's not worth documenting, but I'm okay with leaving the note here.

Comment (sfc-gh-azhan) [reply CHANGELOG.md:78]: since it's a public API, it's safer to document any changes related to it. I agree that we won't find any user to be affected by this.

File: src/snowflake/snowpark/modin/plugin/extensions/index.py (hunk 1 above)
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py:102]: oh, that should be a bug. Good catch! Let me see if I can fix it here.

File: tests/integ/modin/index/test_index_methods.py [hunk 3]
```diff
@@ -359,6 +359,9 @@ def test_index_parent():
     assert_frame_equal(snow_idx._parent, df)
     assert_index_equal(snow_idx, native_idx1)
 
+    with pytest.raises(ValueError):
+        pd.Index(df)
```
Comment (sfc-gh-mvashishtha) [inline tests/integ/modin/index/test_index_methods.py:363]: nit: This assertion seems unrelated to this test case, "Check whether the parent field in Index is updated properly." Can we move it to a separate test case?

**Review by sfc-gh-mvashishtha — APPROVED** [review]

**Review by sfc-gh-nkrishna — APPROVED** [review]
LGTM!
